src/main/kotlin/com/wink/dbse/entity/game/BangCache.kt

```diff
@@ -0,0 +1,234 @@
+    private val userRepository: UserRepository,
+    private val jda: JDA
+) {
+    private val excitementEmote: Emote? = jda.getEmoteById(emoteIds.excitement ?: "")
+    private val bangChannel: TextChannel = jda.getTextChannelById(channelIds.spamChannel ?:
+    throw IllegalStateException("bangChannel not specified")
+    )!!
+    private val threshold: Long = 12000
+    private var timer: Timer? = null
+    private val queue = LinkedList<BangEntity>()
+    private val last20 = LinkedList<BangEntity>()
+    final var panic = false
+        private set
+    final var reset = false
+        private set
+
+    fun add(bangEntity: BangEntity, chambers: Int) {
+        queue.add(bangEntity)
+        last20.add(bangEntity)
+        if (last20.size > 20) {
+            last20.remove(last20.first())
+        }
+
+        checkPanic()
+        if (!panic) {
+            printBangResult(bangEntity, chambers)
+            performBangForResult(bangEntity)
+        }
+    }
+
+    private fun checkPanic() {
+        val avgTime: Long = last20
+            .mapNotNull { it.timeOccurred?.toEpochSecond(ZoneOffset.UTC)?.times(1000) }
+            .reduce { a, b -> a + b }
+            .div(last20.size)
+
+        panic = (avgTime > Date().time - threshold && last20.size >= 20)  || timer != null || reset
+
+        when {
+            timer != null -> {
+                timer!!.cancel()
+                initTimer()
+            }
+            panic -> initTimer()
+            else -> flush()
+        }
+    }
+
+    private fun initTimer() {
+        timer = Timer()
+        timer!!.schedule(initTask(), 1000 * 10, 1000)
+    }
+
+    private fun initTask(): TimerTask {
+        return object : TimerTask() {
+            //Kills the current timer task and flushes the queue
+            override fun run() {
+                if (!reset) finishTimer()
+            }
+        }
+    }
+
+    private fun finishTimer () {
+        timer?.cancel()
+        timer = null
+        manageQueue()
+        panic = false
+    }
+
+    @Scheduled(cron = "59 59 23 * * ?", zone = "GMT-4:00")
+    private fun resetTables () {
+        logger.info("Updating the Bang streak for the day")
+
+        if(panic && timer != null) {
+            reset = true
+            finishTimer()
+            reset = false
+            initTimer()
+        }
+
+        val yesterday = yesterdayBangerRepository.findAll()
+        val today = todayBangerRepository.findAll().mapNotNull { YesterdayBangerEntity(it.userId) }
+        clearDailyTables()
+        updateUserStreaks(yesterday, today)
+        logger.info("Completed Migrating the tables")
+    }
+
+    private fun manageQueue () {
+        if (queue.isEmpty()) {
+            return
+        }
+        val updates = HashMap<Long, LinkedList<BangEntity>>()
+        queue.forEach { updates.getOrPut(it.userId!!) { LinkedList() }.add(it) }
+
+        printResults(updates)
+        performResultUpdates(updates)
```

Comment:
does the action after printing as to ensure it is printed based on the actions then updates them